symbol/symbol.go

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2018 De-labtory
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package symbol
+
+import (
+	"fmt"
+)
+
+type SymbolType string
+
+const (
+	IntegerSymbol  = "INTEGER"
+	BooleanSymbol  = "BOOLEAN"
+	StringSymbol   = "STRING"
+	FunctionSymbol = "FUNCTION"
+)
+
+type Symbol interface {
+	Type() SymbolType
+	String() string
+}
+
+// Represent Integer symbol
+type Integer struct {
+	Value int64
```

Comment:
we don't need to save value, just integer variable name.
for example
```
int a = 1
```
in this case we need to save **'a' and the fact that 'a' is integer**, so `Integer` struct would be like this
```
type Integer struct {
    id Identifier
}

// usage
Integer{id: "a"}
```

Reply:
But in case of using a in other line, how can we inspect value of a?
i.e) int a = 1
a = a + 1

We have symbol table, so we can know a is integer.
But, we don't know value of a.
And all literals defined in ast package, have only their value except their name.
so I think we must have value of symbol in symbol table.

Reply:
Next, I think we don't need name value, because we use name as key of map.
like, map["a"] -> integer a

Reply:
> We have symbol table, so we can know a is integer.
But, we don't know value of a.
And all literals defined in ast package, have only their value except their name.
so I think we must have value of symbol in symbol table.

We don't need to know what value of 'a' is. just need to know what is identifier, what is expression. 
In the case of `a = a + 1`, what `Scope` should do is checking whether lhs `a` is defined before. 
* If not, return error. 
* If defined before, then parsing function builds AST as before (and this is parser part, not `Scope` part)

Reply:
> Next, I think we don't need name value, because we use name as key of map.
like, map["a"] -> integer a

Yes `name` can be redundant, but we use it later after retrieving symbol then validates. for example:

```
obj := scope.Get[token.Val]
if token.Val != obj.name { // we can explicitly compare with name
    ...
}
```

I think it is a little bit style issue :-)